### PR TITLE
Data catalog base and client code generated through OpenAPI

### PR DESCRIPTION
**Pull request by ritwikc**
Signed-off-by: ritwikc <charitwi@in.ibm.com>

This PR address the issue #972
This PR contains:

1. Here data catalog base code generated through openAPI and is placed in `./pkg/taxonomy/model/datacatalog/base`
2. Generated client code is placed in `./pkg/connectors/datacatalog/openapiclient`

### CODE REVIEW

File: charts/fybrik/files/taxonomy/taxonomy.json [hunk 1]
```diff
@@ -46,7 +46,7 @@
           "$ref": "#/definitions/Protocol"
         }
```
Comment (rohithdv) [inline charts/fybrik/files/taxonomy/taxonomy.json:47]: better to not checkin `taxonomy.json` as it is a dynamic file and its contents will depend on what is used to generate it (base + layers).

File: pkg/connectors/datacatalog/openapiclient/.gitignore [hunk 2]
```diff
@@ -0,0 +1,24 @@
+# Compiled Object files, Static and Dynamic libs (Shared Objects)
+*.o
+*.a
+*.so
+
```
Comment (rohithdv) [inline pkg/connectors/datacatalog/openapiclient/.gitignore:5]: no need of this file.

File: pkg/connectors/datacatalog/openapiclient/.openapi-generator-ignore [hunk 3]
```diff
@@ -0,0 +1,23 @@
+# OpenAPI Generator Ignore
+# Generated by openapi-generator https://github.com/openapitools/openapi-generator
+
+# Use this file to prevent files from being overwritten by the generator.
+# The patterns follow closely to .gitignore or .dockerignore.
```
Comment (rohithdv) [inline pkg/connectors/datacatalog/openapiclient/.openapi-generator-ignore:5]: same as above.

File: pkg/connectors/datacatalog/openapiclient/.openapi-generator/FILES [hunk 4]
```diff
@@ -0,0 +1,13 @@
+.gitignore
+.openapi-generator-ignore
+.travis.yml
+README.md
+api/openapi.yaml
```
Comment (rohithdv) [inline pkg/connectors/datacatalog/openapiclient/.openapi-generator/FILES:5]: this is also not needed. We generate openapiclient only once unlike base which can be generated multiple times. So it is better to remove unnecessary files created during code generation. Thanks.

**Review by rohithdv — APPROVED** [review]
Suggested minor changes. Please change them. I am fine with other changes. Thanks.

File: pkg/connectors/datacatalog/openapiclient/.gitignore (hunk 2 above)
Comment (ritwikc) [reply pkg/connectors/datacatalog/openapiclient/.gitignore]: Done

File: pkg/connectors/datacatalog/openapiclient/.openapi-generator-ignore (hunk 3 above)
Comment (ritwikc) [reply pkg/connectors/datacatalog/openapiclient/.openapi-generator-ignore]: Done

File: pkg/connectors/datacatalog/openapiclient/.openapi-generator/FILES (hunk 4 above)
Comment (ritwikc) [reply pkg/connectors/datacatalog/openapiclient/.openapi-generator/FILES]: Done

File: charts/fybrik/files/taxonomy/taxonomy.json (hunk 1 above)
Comment (ritwikc) [reply charts/fybrik/files/taxonomy/taxonomy.json:47]: I have made the file at par with the current version in master repository.

Comment (ritwikc) [reply charts/fybrik/files/taxonomy/taxonomy.json]: Rebased the code.

File: .github/workflows/golang-security.yml [hunk 5]
```diff
@@ -16,4 +16,4 @@ jobs:
       - name: Run Gosec Security Scanner
         uses: securego/gosec@master
         with:
-          args: -exclude-dir=pkg/connectors/policymanager/openapiclient -exclude-dir=connectors/opa -exclude-dir=cmd -exclude-dir=test/services/pilot/mock-openapi-policymanager ./... 
+          args: -exclude-dir=pkg/connectors/policymanager/openapiclient -exclude-dir=pkg/connectors/datacatalog/openapiclient -exclude-dir=connectors/opa -exclude-dir=cmd -exclude-dir=test/services/pilot/mock-openapi-policymanager ./...
```
[the remark this answers is not in the document]
Comment (ritwikc) [reply .github/workflows/golang-security.yml:19]: The generated code does not pass go security. That is why it is excluded.

File: charts/fybrik/files/taxonomy/taxonomy.json [hunk 6]
```diff
@@ -46,7 +46,7 @@
...
       },
-      "additionalProperties": false
+      "additionalProperties": true
```
[the remark this answers is not in the document]
Comment (ritwikc) [reply charts/fybrik/files/taxonomy/taxonomy.json]: Rebased the code.

File: .github/workflows/golang-security.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (ronenkat) [reply .github/workflows/golang-security.yml:19]: @ritwikc please share what are the security issues that are in the generated code.
If the generated code pose a risk, we will need to find an alternative.

[the remark this answers is not in the document]
Comment (rohithdv) [reply .github/workflows/golang-security.yml:19]: @ronenkat Actually we are excluding security checks for other folders in fybrik as given [here](https://github.com/fybrik/fybrik/blob/d18b0ddc6d3c3c67573e3bc91c2ca598e1109879/.github/workflows/golang-security.yml#L19). The decision made with @roee88 was to add the generated code to the exclusion list until we complete the implementation of the taxonomy work. 

Probably @ritwikc can you paste the errors we get if we dont add to the exclusion list here for reference?

[the remark this answers is not in the document]
Comment (ritwikc) [reply .github/workflows/golang-security.yml:19]: This is the error. @ronenkat @rohithdv 
```
[/github/workspace/pkg/connectors/datacatalog/openapiclient/client.go:400] - G304 (CWE-22): Potential file inclusion via variable (Confidence: HIGH, Severity: MEDIUM)
    399: func addFile(w *multipart.Writer, fieldName, path string) error {
  > 400: 	file, err := os.Open(path)
    401: 	if err != nil {



[/github/workspace/pkg/connectors/datacatalog/openapiclient/client.go:404] - G307 (CWE-703): Deferring unsafe method "Close" on type "*os.File" (Confidence: HIGH, Severity: MEDIUM)
    403: 	}
  > 404: 	defer file.Close()
    405: 



[/github/workspace/pkg/connectors/datacatalog/openapiclient/client.go:258] - G104 (CWE-703): Errors unhandled. (Confidence: HIGH, Severity: LOW)
    257: 		headerParams["Content-Length"] = fmt.Sprintf("%d", body.Len())
  > 258: 		w.Close()
    259: 	}



[/github/workspace/pkg/connectors/datacatalog/openapiclient/client.go:236] - G104 (CWE-703): Errors unhandled. (Confidence: HIGH, Severity: LOW)
    235: 				} else { // form value
  > 236: 					w.WriteField(k, iv)
    237: 				}



[/github/workspace/pkg/connectors/datacatalog/openapiclient/api_default.go:123] - G104 (CWE-703): Errors unhandled. (Confidence: HIGH, Severity: LOW)
    122: 	localVarBody, err := _ioutil.ReadAll(localVarHTTPResponse.Body)
  > 123: 	localVarHTTPResponse.Body.Close()
    124: 	localVarHTTPResponse.Body = _ioutil.NopCloser(bytes.NewBuffer(localVarBody))



Summary:
  Gosec  : v2.9.3
  Files  : 130
  Lines  : 19213
  Nosec  : 2
  Issues : 5
```
The entire go security scanner result is in [here](https://github.com/ritwikc/the-mesh-for-data/runs/4344680530?check_suite_focus=true)

**Review by ronenkat — APPROVED** [review]
lets keep in mind that we need to handle the security issues in the future.

**Pull request merged**